### PR TITLE
[runtime] AOTExecutor implementation and c target code-generator

### PR DESCRIPTION
**Pull request by areusch**
This PR adds the AOTExecutor implementation and the metadata code-generator for use with the `c` target. This PR allows users to run models which use only the CPU via the AOTExecutor using the C++ runtime.

cc @Mousius @manupa-arm @kparzysz-quic @masahi @mehrdadh

### CODE REVIEW

File: tests/python/relay/aot/test_cpp_aot.py
Comment (masahi) [inline tests/python/relay/aot/test_cpp_aot.py:51]: need to remove llvm for now

Comment (areusch) [reply tests/python/relay/aot/test_cpp_aot.py:51]: done

**Comment by areusch** [comment]
@Mousius @manupa-arm embarrassingly i can't quite figure out how to reproduce the Ethos-U test failures locally:
```
$ docker/bash.sh -it ci_qemu bash -c 'cd build && cmake -GNinja .. && ninja && cd .. && tests/scripts/task_ci_s
etup.sh && . tests/scripts/setup-pytest-env.sh  && python3  -mpytest tests/python/contrib/test_ethosu -k "test_tflite_squeeze[ifm_shape1-3-ethos-
u55-256]"'

...
==================================================================== ERRORS =====================================================================
_________________________________________________________ ERROR collecting test session _________________________________________________________
/usr/lib/python3.6/importlib/__init__.py:126: in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
<frozen importlib._bootstrap>:994: in _gcd_import
    ???
<frozen importlib._bootstrap>:971: in _find_and_load
    ???
<frozen importlib._bootstrap>:955: in _find_and_load_unlocked
    ???
<frozen importlib._bootstrap>:665: in _load_unlocked
    ???
/opt/tvm-venv/lib/python3.6/site-packages/_pytest/assertion/rewrite.py:170: in exec_module
    exec(co, module.__dict__)
tests/python/contrib/test_ethosu/cascader/conftest.py:49: in <module>
    from ..infra import (
tests/python/contrib/test_ethosu/infra.py:44: in <module>
    from tests.python.relay.aot.aot_test_utils import (
E   ModuleNotFoundError: No module named 'tests.python'
```

I can't find where we add tests as a valid import path. I have a vague recollection we discussed this but can't find where it was. do you know? hoping the CI passes if i fixed the unittests.

**Comment by masahi** [comment]
@areusch Could the ethos-u failure be a flaky one? We had a lot of them recently.

**Comment by manupa-arm** [comment]
The above sounds like a import issue with tests modules.

@masahi is it the same issue you were seeing? If so it shouldn't not be flaky rather it should fail always.

I can have a look tommorow but unfortunately haven't encountered it personally.

**Comment by masahi** [comment]
No this one https://ci.tlcpack.ai/blue/organizations/jenkins/tvm/detail/PR-10283/17/pipeline/307

**Comment by manupa-arm** [comment]
Seems like some u65 tests are failing. Please can you file an issue?

cc : @ekalda

**Comment by masahi** [comment]
I think there is already https://github.com/apache/tvm/issues/10300 but not sure if it is the same one.

The point is we have many flaky ethos-u issues going on lately.

**Comment by manupa-arm** [comment]
I would expect to see something here at least : https://ci.tlcpack.ai/blue/organizations/jenkins/tvm/activity/?branch=main.

I think as part of that issue the flaky ones were disabled. @lhutton1 is looking into those.  I think its worth mentioning if you see more at least in that issue, that would aid in triaging these...

**Comment by areusch** [comment]
if we see more of the same flakiness with similar traceback we can add it to #10300; I'm not convinced it's the same thing here just yet (since other unittests were failing and that was due to a divergence in the way `tvm.build` and `tvm.relay.build` populate IRModule attrs (see #10373 ).

**Comment by ekalda** [comment]
Yeah the Ethos-U test have been really flaky lately, there is always a random assortment failing and a random assortment has been disabled. By the looks of it, there is something wrong with running the FVP, @lhutton1 has been looking into it, but we can't reproduce it locally, so we would need to progress https://github.com/apache/tvm/pull/10214 to get a somewhat more meaningful errors in upstream CI...

File: src/relay/backend/aot_executor_codegen.cc
Comment (kparzysz-quic) [inline src/relay/backend/aot_executor_codegen.cc:187]: All of this is equivalent to `return_ttypes_ = FlattenTupleType(e->checked_type());`.

**Review by kparzysz-quic — APPROVED** [review]
Looks ok to me.  Please change the `make_unique` statements though, but it can go into a subsequent PR.

File: src/target/source/source_module.cc
Comment (masahi) [inline src/target/source/source_module.cc:557]: Remove

File: src/target/source/source_module.cc
Comment (masahi) [inline src/target/source/source_module.cc:693]: Remove

File: src/target/source/source_module.cc
Comment (masahi) [inline src/target/source/source_module.cc:666]: Remove

File: src/target/source/source_module.cc
Comment (masahi) [inline src/target/source/source_module.cc:562]: Remove

File: src/target/source/source_module.cc
Comment (masahi) [inline src/target/source/source_module.cc:728]: Remove

**Review by masahi — COMMENTED** [review]
Found some left-over log while running `test_conv2d`

File: src/target/metadata_module.cc
Comment (masahi) [inline src/target/metadata_module.cc:110]: Check that `metadata->pool_inputs` is non-null (it's `Optional<Map<tir::Var, tir::usmp::AllocatedPoolInfo>>`)

File: python/tvm/runtime/executor/aot_executor.py
Comment (masahi) [inline python/tvm/runtime/executor/aot_executor.py:55]: Need update?

File: python/tvm/runtime/executor/__init__.py
Comment (masahi) [inline python/tvm/runtime/executor/__init__.py:24]: I just realized that, we have two notions of `executor`. One is runtime one above, the other is 
https://github.com/apache/tvm/blob/55849e651e3dff6cffeb7fd14aa89699e9575d10/python/tvm/relay/build_module.py#L647

Do we intend to support `create_executor(kind="aot", ...)`?

File: include/tvm/runtime/metadata.h
Comment (masahi) [inline include/tvm/runtime/metadata.h:78]: You are not using `run_model` anymore, are you? In your previous branch indeed I'm seeing `tvmgen_default_run_model` generated, but after rebase it is replaced with `tvmgen_default___tvm_main__`.

Comment (manupa-arm) [reply include/tvm/runtime/metadata.h:78]: That is right!.

We had run (entry_point), run_model and  __tvm_main__.
The entry_point is supposed to call run_model, however, run_model is identical to tvm_main -- therefore, it was removed due to no need of maintaining two symbols for relay and tir versions of main.

So I think it needs to be tvm_main now.

File: python/tvm/runtime/executor/aot_executor.py
Comment (manupa-arm) [inline python/tvm/runtime/executor/aot_executor.py:22]: (stylistic) : AotModule --> AOTModule or AoTModule ?

File: src/relay/backend/aot_executor_codegen.cc
Comment (manupa-arm) [inline src/relay/backend/aot_executor_codegen.cc:330]: nit : any reason mixed pass-by-value, pass-by-(const)-references usage and pointers here ?

File: src/relay/backend/aot_executor_codegen.cc
Comment (manupa-arm) [inline src/relay/backend/aot_executor_codegen.cc:918]: Are both of these CHECKs tested ?

File: src/runtime/aot_executor/aot_executor_factory.cc
Comment (manupa-arm) [inline src/runtime/aot_executor/aot_executor_factory.cc:22]: Need update

File: src/runtime/aot_executor/aot_executor_factory.cc
Comment (manupa-arm) [inline src/runtime/aot_executor/aot_executor_factory.cc:37]: Just for my knowledge, I never understood why these are called "factories". What maybe the reason ?  :)

File: tests/python/relay/aot/test_crt_aot.py
Comment (manupa-arm) [inline tests/python/relay/aot/test_crt_aot.py:55]: From the above comment, I'd expect two of these tests here.

**Review by manupa-arm — COMMENTED** [review]
Neat! Looking good.

Few nits and questions (for my understanding).

File: src/target/source/source_module.cc
Comment (masahi) [inline src/target/source/source_module.cc:678]: Remove?

File: src/target/source/source_module.cc
Comment (masahi) [inline src/target/source/source_module.cc:814]: Remove?

File: src/target/source/source_module.cc
Comment (masahi) [inline src/target/source/source_module.cc:813]: Rather than hard code `get_c_metadata` everywhere, we should introduce a new symbol at https://github.com/apache/tvm/blob/d8f639a84fd9a11f03a949eb751a18018a9c1033/include/tvm/runtime/module.h#L219 (similar to `tvm_module_main`)

File: src/tir/transforms/legalize_packed_calls.cc
Comment (masahi) [inline src/tir/transforms/legalize_packed_calls.cc:83]: What are these changes for?

**Review by masahi — COMMENTED** [review]
More minor comments, finished the whole thing.

At some point we should review our meta data situation... There are some many classes / files and they look quite messy.

File: src/runtime/metadata.cc
Comment (masahi) [inline src/runtime/metadata.cc:90]: See https://github.com/apache/tvm/pull/10283/files#r817105544

**Review by areusch — COMMENTED** [review]
@manupa-arm @kparzysz-quic @masahi thanks for the review! addressed your comments, ptal when you have time.

File: src/target/metadata_module.cc
Comment (masahi) [reply src/target/metadata_module.cc:112]: Is it done?

File: src/relay/backend/aot_executor_codegen.cc
Comment (masahi) [reply src/relay/backend/aot_executor_codegen.cc:187]: cc @areusch

**Review by masahi — APPROVED** [review]
LGTM modulo some unaddressed comments. But they can go into the next LLVM PR.

https://github.com/apache/tvm/pull/10283#discussion_r818004670
https://github.com/apache/tvm/pull/10283#discussion_r816657318

**Comment by areusch** [comment]
@manupa-arm @kparzysz-quic please take a look when you have a minute, CI is green now.

**Review by tmoreau89 — APPROVED** [review]
LGTM - thanks @areusch for landing AoT executor !

**Pull request merged**
